src/invidious/helpers/sig_helper.cr

```diff
@@ -0,0 +1,325 @@
+
+  abstract struct Payload
+  end
+
+  struct StringPayload < Payload
+    getter string : String
+
+    def initialize(str : String)
+      raise Exception.new("SigHelper: String can't be empty") if str.empty?
+      @string = str
+    end
+
+    def self.from_bytes(slice : Bytes)
+      size = IO::ByteFormat::NetworkEndian.decode(UInt16, slice)
+      if size == 0 # Error code
+        raise Exception.new("SigHelper: Server encountered an error")
+      end
+
+      if (slice.bytesize - 2) != size
+        raise Exception.new("SigHelper: String size mismatch")
+      end
+
+      if str = String.new(slice[2..])
+        return self.new(str)
+      else
+        raise Exception.new("SigHelper: Can't read string from socket")
+      end
+    end
+
+    def to_io(io)
+      # `.to_u16` raises if there is an overflow during the conversion
+      io.write_bytes(@string.bytesize.to_u16, NetworkEndian)
+      io.write(@string.to_slice)
+    end
+  end
+
+  private enum Opcode
+    FORCE_UPDATE            = 0
+    DECRYPT_N_SIGNATURE     = 1
+    DECRYPT_SIGNATURE       = 2
+    GET_SIGNATURE_TIMESTAMP = 3
+    GET_PLAYER_STATUS       = 4
+  end
+
+  private record Request,
+    opcode : Opcode,
+    payload : Payload?
+
+  # ----------------------
+  #  High-level functions
+  # ----------------------
+
+  module Client
+    extend self
+
+    # Forces the server to re-fetch the YouTube player, and extract the necessary
+    # components from it (nsig function code, sig function code, signature timestamp).
+    def force_update : UpdateStatus
+      request = Request.new(Opcode::FORCE_UPDATE, nil)
+
+      value = send_request(request) do |bytes|
+        IO::ByteFormat::NetworkEndian.decode(UInt16, bytes)
+      end
+
+      case value
+      when 0x0000 then return UpdateStatus::Error
+      when 0xFFFF then return UpdateStatus::UpdateNotRequired
+      when 0xF44F then return UpdateStatus::Updated
+      else
+        code = value.nil? ? "nil" : value.to_s(base: 16)
+        raise Exception.new("SigHelper: Invalid status code received #{code}")
+      end
+    end
+
+    # Decrypt a provided n signature using the server's current nsig function
+    # code, and return the result (or an error).
+    def decrypt_n_param(n : String) : String?
+      request = Request.new(Opcode::DECRYPT_N_SIGNATURE, StringPayload.new(n))
+
+      n_dec = send_request(request) do |bytes|
+        StringPayload.from_bytes(bytes).string
+      end
+
+      return n_dec
+    end
+
+    # Decrypt a provided s signature using the server's current sig function
+    # code, and return the result (or an error).
+    def decrypt_sig(sig : String) : String?
+      request = Request.new(Opcode::DECRYPT_SIGNATURE, StringPayload.new(sig))
+
+      sig_dec = send_request(request) do |bytes|
+        StringPayload.from_bytes(bytes).string
+      end
+
+      return sig_dec
+    end
+
+    # Return the signature timestamp from the server's current player
+    def get_sts : UInt64?
```

Comment:
Can you add an `ameba:disable Naming/AccessorMethodName` to the definitions of `get_*`? Ameba seems to be mistakenly classifying these as getter methods.

Reply:
I'd rather disable `Naming/AccessorMethodName` altogether. Maybe `Naming/BlockParameterName` too!

Reply:
Disabling `Naming/BlockParameterName` is fine imo but the general conventions in Crystal is to suffix accessor  methods with `?`

Shouldn't Invidious be adhering to that convention? I'm not sure if the amount of false positives within Ameba is enough to warrant disabling that rule altogether.

Reply:
I take back what I said. Ameba complains about `Naming/AccessorMethodName` as a false positive more often than not